take-five.js

```diff
@@ -1,64 +1,238 @@
-    if (!routers.has(method)) {
-      routers.set(method, wayfarer('/_'))
+  _handleRequest (req, res, ctx) {
+    try {
+      const method = req.method.toLowerCase()
+      const url = req.url.split('?')[0]
+      const router = this.routers.get(method)
+      router(url, req, res, ctx)
+    } catch (err) {
+      console.log('am i here')
+      if (res.finished) {
+        throw err
+      }
+      return ctx.err(404, 'Not found')
     }
+  }
+
+  _verifyBody (req, res, ctx) {
+    const type = req.headers['content-type']
+    const size = req.headers['content-length']
 
-    if (!Array.isArray(funcs)) {
-      funcs = [funcs]
+    if (type !== 'application/json') {
+      return ctx.err(415, `POST requests must be application/json not ${type}`)
     }
 
-    routers.get(method).on(matcher, (params, req, res) => {
-      req.params = querystring.parse(req.url.split('?')[1])
-      req.urlParams = params
-      handleRequest(req, res, funcs)
+    if (size > this.maxPost) {
+      return ctx.err(413, `${size} exceeds maximum size for requests`)
+    }
+
+    const parser = concat((data) => {
+      if (!res.finished) {
+        ctx.body = data
+        this.parseBody(req, res, ctx)
+      }
+    })
+    req.pipe(parser)
+
+    const body = []
+    req.on('data', (chunk) => {
+      if (chunk.length > this.maxPost || Buffer.byteLength(body.join(''), 'utf8') > this.maxPost) {
+        return ctx.err(413, 'Payload size exceeds maxmium body length')
+      }
     })
   }
+
+  _onRequest (req, res) {
+    this.cors(res)
+
+    if (req.method === 'OPTIONS') {
+      res.statusCode = 203
+      return res.end()
+    }
+
+    const ctx = this.makeCtx(res)
+    const conlen = parseInt(req.headers['content-length'], 10) || 0
+
+    if (conlen !== 0 && dataMethods.includes(req.method.toLowerCase())) {
+      this._verifyBody(req, res, ctx)
+    } else {
+      this._handleRequest(req, res, ctx)
+    }
+  }
+
+  _addRouters () {
+    this.methods.forEach((method) => {
+      Object.defineProperty(this, method, {value: generateRouter(method)})
+    })
+
+    function generateRouter (method) {
+      return function (matcher, handler) {
+        let router = this.routers.get(method)
+        if (!router) {
+          router = wayfarer('/_')
+          this.routers.set(method, router)
+        }
+
+        const handlers = Array.isArray(handler) ? handler : [handler]
+
+        if (handlers.some((f) => typeof f !== 'function')) {
+          throw new Error('handlers must be functions')
+        }
+
+        router.on(matcher, (params, req, res, ctx) => {
+          ctx.query = querystring.parse(req.url.split('?')[1])
+          ctx.params = params
+          this._resolveHandlers(req, res, ctx, handlers.slice(0))
+        })
+      }
+    }
+  }
+
+  _resolveHandlers (req, res, ctx, handlers) {
+    const next = handlers.shift()
+    iterate(next)
+
+    function iterate (handler) {
+      const done = (value) => {
```

Comment:
@ungoldman i struggled with this part -- the problem with promises is that they're not deferred; so you can't like "pass a promise" around, so using `Promise.race` didn't work since I couldn't get the promise back out of the `ctx.next` function.

Then I realized that I could just pass in the `then` handler for the promise as `ctx.next`.  Wondering if I should attempt to guard against the user both resolving a promise AND making the callback, but not sure if that's my responsibility?

Reply:
If it's just a one or two line addition I think it's worth surfacing the error to the user.

Reply:
No longer allowing function continuation passing -- all methods must be async.